### PR TITLE
Add catalog support

### PR DESCRIPTION
**Pull request by aclevername**
### Description

Adds support for `ProfleCatalogSources`. When a catalog resource gets added it gets stored in-memory in an API. You can query this API to search through the catalogs to find matching profiles. Intended to be used incombination with pctl https://github.com/weaveworks/pctl/pull/3

### Checklist
- [ ] Added tests that cover your change
- [ ] Added/modified documentation as required (such as the `README.md`)
- [ ] Made sure the title of the PR is a good description that can go into the release notes
- [ ] Added a `kind` label to the PR (e.g. `kind/feature`)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -8,6 +8,7 @@ Gitops native package management
     the `profilesubscription`, `source` and `helm` controllers.
 
 1. Then subscribe to the example [nginx-profile](https://github.com/weaveworks/nginx-profile): `kubectl apply -f examples/profile-subscription.yaml`
+and deploy an example catalog `kubectl apply -f examples/profile-catalog-source.yaml`
```
Comment (Callisto13) [inline README.md:11]: can we change the ordering of the steps here?

i think they should be something like:

1. deploy alll crds
2. create catalog
3. look at catalog
4. subscribe
5. profit

File: api/v1alpha1/profilecatalogsource_types.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// ProfileCatalogSourceSpec defines the desired state of ProfileCatalogSource
+type ProfileCatalogSourceSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
```
Comment (Callisto13) [inline api/v1alpha1/profilecatalogsource_types.go:28]: lots of boilerplate notes left in this file

File: config/samples/_v1alpha1_profilecatalogsource.yaml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+apiVersion: weave.works/v1alpha1
```
Comment (Callisto13) [inline config/samples/_v1alpha1_profilecatalogsource.yaml:1]: did we decide to remove these samples in favour of those in `examples`?

File: controllers/profilecatalogsource_controller.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the ProfileCatalogSource object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.7.0/pkg/reconcile
```
Comment (Callisto13) [inline controllers/profilecatalogsource_controller.go:52]:
```suggestion
```
boilerplate

File: controllers/suite_test.go [hunk 5]
```diff
@@ -34,13 +34,18 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/log/zap"
 
 	"github.com/weaveworks/profiles/api/v1alpha1"
+	weaveworksv1alpha1 "github.com/weaveworks/profiles/api/v1alpha1"
```
Comment (Callisto13) [inline controllers/suite_test.go:37]: dupe

File: main.go [hunk 6]
```diff
@@ -54,11 +60,12 @@ func init() {
 }
 
 func main() {
-	var metricsAddr string
 	var enableLeaderElection bool
-	var probeAddr string
+	var metricsAddr, probeAddr, apiAddr string
 	flag.StringVar(&metricsAddr, "metrics-bind-address", ":8080", "The address the metric endpoint binds to.")
 	flag.StringVar(&probeAddr, "health-probe-bind-address", ":8081", "The address the probe endpoint binds to.")
+	flag.StringVar(&apiAddr, "profiles-api-bind-address", ":8000", "The address the profiles api binds to.")
```
Comment (Callisto13) [inline main.go:67]:
```suggestion
	flag.StringVar(&apiAddr, "profiles-api-bind-address", ":8000", "The address the profiles catalog api binds to.")
```
just to be super clear

File: main.go [hunk 7]
```diff
@@ -103,6 +121,15 @@ func main() {
 		os.Exit(1)
 	}
 
+	setupLog.Info(fmt.Sprintf("starting profiles api server at %s", apiAddr))
+	go func() {
+		err := http.ListenAndServe(apiAddr, api.New(profileCatalog))
```
Comment (Callisto13) [inline main.go:126]: we probably want some graceful shutdown handling at some point

**Review by Callisto13 — APPROVED** [review]
Not thrilled with the name `api` for that package. But I do need this merged now

File: README.md (hunk 1 above)
Comment (aclevername) [reply README.md:11]: done

**Pull request merged**